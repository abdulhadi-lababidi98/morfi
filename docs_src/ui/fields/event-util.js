/**
 * This file is part of morfi which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

type Event = { preventDefault: Function, target: { value: any } };
type EventHandler = Event => void;

type CustomInputHandling = { preventDefault?: boolean, falsyToUndefined?: boolean };

const inputHandler = (
    cb: any => mixed | void,
    { preventDefault = true, falsyToUndefined = false }: CustomInputHandling = {}
): EventHandler => {
    return event => {
        if (preventDefault) event.preventDefault();
        const original = event.target.value;
        cb(falsyToUndefined ? original || undefined : original);
    };
};

export const EventUtil = { inputHandler };
