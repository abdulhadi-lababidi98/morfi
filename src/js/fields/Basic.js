/**
 * This file is part of morfi which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import React from 'react';
import type { ErrorMessage } from '../form';
import messages from '../../messages';

export const __ = ({ id, values }: ErrorMessage): string => {
    let result = messages[id];
    if (result && values) {
        result = Object.keys(values).reduce(
            (red: string, key: string) => red.replace(`{${key}}`, String(values[key])),
            result
        );
    }
    return result;
};

export const Label = ({ label, required = false }: { label: string, required?: boolean }) => {
    return <label className="control-label">{label + (required ? ' *' : '')}</label>;
};

export const DisplayError = ({ error }: { error: ErrorMessage }) => {
    return <span className="invalid-feedback">{__(error)}</span>;
};

type EventHandler = (SyntheticInputEvent<*>) => void;

export const onActionWrap = (cb?: any => void, preventDefault: boolean = true): EventHandler => {
    return event => {
        if (preventDefault) {
            event.preventDefault();
        }
        if (cb) {
            cb(event.target.value);
        }
    };
};
