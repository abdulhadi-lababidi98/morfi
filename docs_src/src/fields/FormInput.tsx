import React, { Component } from 'react';
import { DisplayError, Label, onActionWrap } from './Basic';
import { Eye } from '../icons/Eye';
import { Spinner } from '../icons/Spinner';
import { Morfi, FieldControls, FormField } from '../../../src';

type AllowedTypes = 'text' | 'password';

type CommonInputProps = {
    label?: string;
    className?: string;
    autoFocus?: boolean;
    disabled?: boolean;
    type?: AllowedTypes;
    placeholder?: string;
    pending?: boolean;
};

type InputProps = FieldControls<string> & CommonInputProps;

export class Input extends Component<InputProps, { showPassword: boolean }> {
    state = { showPassword: false };
    showPasswordToggle = () => this.setState((state) => ({ showPassword: !state.showPassword }));
    type = () => (this.props.type !== 'password' ? this.props.type : this.state.showPassword ? 'text' : 'password');
    render() {
        const {
            value = '',
            label,
            error,
            required,
            className = 'form-group',
            onChange,
            onBlur,
            autoFocus,
            disabled,
            type = 'text',
            placeholder,
            pending,
        } = this.props;
        return (
            <div className={className}>
                {label && <Label {...{ label, required }} />}
                <input
                    onChange={onActionWrap(onChange)}
                    onBlur={onActionWrap(onBlur)}
                    disabled={disabled}
                    className={'form-control' + (error ? ' is-invalid' : '')}
                    type={this.type()}
                    {...{ placeholder, value, autoFocus }}
                />
                {type === 'password' && (
                    <span onClick={this.showPasswordToggle}>
                        <Eye stroked={this.state.showPassword} />
                    </span>
                )}
                {pending && <Spinner />}
                {error && <DisplayError error={error} />}
            </div>
        );
    }
}

type FormInputProps = { field: FormField<string> } & CommonInputProps;

export const FormInput: React.FC<FormInputProps> = ({ field, ...rest }) => {
    const fieldProps = Morfi.useField(field);
    return <Input {...fieldProps} {...rest} />;
};
