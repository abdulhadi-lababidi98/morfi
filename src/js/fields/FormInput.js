/**
 * This file is part of morfi which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import React from 'react';
import { Error, Label, onActionWrap } from './Basic';
import { Field } from '../form';
import type { _ErrorMessage } from '../form/Form-classes';
import { Spinner } from '../components/Spinner';

type AllowedTypes = 'text' | 'password';

type InputProps = {
    value?: string,
    label?: string,
    error?: _ErrorMessage,
    required?: boolean,
    className?: string,
    onChange?: string => void,
    onBlur?: string => void,
    autoFocus?: boolean,
    disabled?: boolean,
    type?: AllowedTypes,
    placeholder?: string,
    pending?: boolean,
};

export const Input = ({
    value = '',
    label,
    error,
    required,
    className,
    onChange,
    onBlur,
    autoFocus,
    disabled,
    type = 'text',
    placeholder,
    pending,
}: InputProps) => {
    return (
        <div className={className}>
            {label && <Label {...{ label, required }} />}
            <input
                onChange={onActionWrap(onChange)}
                onBlur={onActionWrap(onBlur)}
                disabled={disabled || pending}
                className={'form-control' + (error ? ' is-invalid' : '')}
                {...{ type, placeholder, value, autoFocus }}
            />
            {pending && <Spinner />}
            {error && <Error error={error} />}
        </div>
    );
};

type FormInputProps = {|
    name: string,
    value?: string,
    error?: _ErrorMessage,
    label?: string,
    className?: string,
    disabled?: boolean,
    autoFocus?: boolean,
    type?: AllowedTypes,
    placeholder?: string,
    pending?: boolean,
|};

export const _FormInput = ({
    name,
    value,
    error,
    label,
    className = 'form-group',
    disabled,
    autoFocus,
    type,
    placeholder,
    pending,
}: FormInputProps) => (
    <Field name={name}>
        {({ onChange, onBlur, required }) => (
            <Input
                {...{
                    error,
                    value,
                    required,
                    onBlur,
                    onChange,
                    label,
                    className,
                    disabled,
                    autoFocus,
                    type,
                    placeholder,
                    pending,
                }}
            />
        )}
    </Field>
);
