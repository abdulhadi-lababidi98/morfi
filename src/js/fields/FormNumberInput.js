/**
 * This file is part of form4react which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import React from 'react';
import { Error, Label, onActionWrap } from './Basic';
import { Field } from '../form';
import type { _ErrorMessage } from '../form/Form-classes';

type NumberInputProps = {
    value?: number,
    label?: string,
    error?: _ErrorMessage,
    required?: boolean,
    className?: string,
    onChange?: number => void,
    onBlur?: number => void,
    autoFocus?: boolean,
    disabled?: boolean,
};

export const NumberInput = ({
    value = 0,
    label,
    error,
    required,
    className,
    onChange,
    onBlur,
    autoFocus,
    disabled,
}: NumberInputProps) => {
    return (
        <div className={(className || '') + (error ? ' has-error' : '')}>
            {label && <Label {...{ label, required }} />}
            <input
                type="number"
                onChange={onActionWrap(onChange)}
                onBlur={onActionWrap(onBlur)}
                className="form-control"
                {...{ value, autoFocus, disabled }}
            />
            {error && <Error error={error} />}
        </div>
    );
};

type FormNumberInputProps = {|
    name: string,
    value?: number,
    error?: _ErrorMessage,
    label?: string,
    className?: string,
    disabled?: boolean,
    autoFocus?: boolean,
|};

export const _FormNumberInput = ({
    name,
    value,
    error,
    label,
    className = 'form-group',
    disabled,
    autoFocus,
}: FormNumberInputProps) => (
    <Field name={name}>
        {({ onChange, onBlur, required }) => (
            <NumberInput {...{ error, value, required, onBlur, onChange, label, className, disabled, autoFocus }} />
        )}
    </Field>
);
