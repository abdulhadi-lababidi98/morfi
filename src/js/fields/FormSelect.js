/**
 * This file is part of form4react which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import React from 'react';
import type { _ErrorMessage } from '../form/Form-classes';
import { _Field as Field } from '../form/Field';
import { Error, Label, onActionWrap } from './Basic';

export type SelectOption = {| label: string, value: string |};

type SelectProps = {
    value?: string,
    label?: string,
    error?: _ErrorMessage,
    required?: boolean,
    className?: string,
    onChange?: string => void,
    onBlur?: string => void,
    options: Array<SelectOption>,
    disabled?: boolean,
};

const Option = ({ option }: { option: SelectOption }) => <option value={option.value}>{option.label}</option>;

const Select = ({ value, label, error, required, className, onChange, onBlur, options, disabled }: SelectProps) => {
    return (
        <div className={className}>
            {label && <Label {...{ label, required }} />}
            <select
                className="form-control"
                disabled={disabled}
                onChange={onActionWrap(onChange)}
                onBlur={onActionWrap(onBlur)}
                value={value}>
                {options.map((option: SelectOption, index: number) => <Option option={option} key={index} />)}
            </select>
            {error && <Error error={error} />}
        </div>
    );
};

type FormSelectProps = {|
    name: string,
    value?: string,
    error?: _ErrorMessage,
    label?: string,
    className?: string,
    options: Array<SelectOption>,
    disabled?: boolean,
|};

export const _FormSelect = ({
    name,
    value,
    error,
    options,
    label,
    className = 'form-group',
    disabled,
}: FormSelectProps) => (
    <Field name={name}>
        {({ onChange, onBlur, required }) => (
            <Select {...{ error, value, required, onBlur, onChange, label, className, disabled, options }} />
        )}
    </Field>
);
