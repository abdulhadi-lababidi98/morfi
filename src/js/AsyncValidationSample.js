/**
 * This file is part of morfi which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import React, { Component } from 'react';
import { Validators } from './validators/validators';
import { Form } from './form';
import { FormInput } from './fields';
import type { FormData, FormValidation, Validator } from './form';

type FormValues = {
    userName?: string,
    realName?: string,
};

const StaticValidations: FormValidation<FormValues> = {
    realName: {
        onChange: Validators.optionalOf(
            Validators.regex({
                re: /^[a-zöüä][a-zöüä ]*$/i,
                message: { id: 'AsyncValidationSample.realName.validation.requirements' },
            })
        ),
    },
};

type AsyncValidationSampleState = {
    data: FormData<FormValues>,
    validations: FormValidation<FormValues>,
    pendingUserName: boolean,
};

const alreadyRegistered_userName = 'tom';
const everySecondTimeFailing_userName = 'jack';
const failingAfterSubmit_userName = 'mike';
let validationCounter = 0;

export default class AsyncValidationSample extends Component<{}, AsyncValidationSampleState> {
    validateUserName: Validator = (userName?: string) => {
        const syncError = Validators.string({ min: 1 })(userName);
        if (syncError) {
            return syncError;
        }
        const lowerCaseUserName = (userName: any).toLowerCase();
        this.setState({ pendingUserName: true });
        validationCounter++;
        return window.sleep(1000).then(() => {
            if (lowerCaseUserName === alreadyRegistered_userName) {
                this.setState({ pendingUserName: false });
                return { id: 'AsyncValidationSample.userName.already.registered', values: { userName } };
            } else if (lowerCaseUserName === everySecondTimeFailing_userName && validationCounter % 2 === 0) {
                this.setState({ pendingUserName: false });
                return { id: 'AsyncValidationSample.userName.already.registered', values: { userName } };
            } else {
                this.setState({ pendingUserName: false });
            }
        });
    };

    getValidations = (): FormValidation<FormValues> => {
        return {
            ...StaticValidations,
            userName: { onBlur: this.validateUserName, onChange: Validators.string({ min: 1 }) },
        };
    };

    state: AsyncValidationSampleState = {
        data: { values: {}, errors: {} },
        pendingUserName: false,
        validations: this.getValidations(),
    };

    onChange = (data: FormData<FormValues>) => this.setState({ data });

    onSubmit = ({ userName = '' }: FormValues): Promise<void> => {
        console.log('onSubmit was called');
        // simulate server request
        const fakeServerRequest = window.sleep(1000);
        return userName.toLowerCase() !== failingAfterSubmit_userName
            ? fakeServerRequest
            : fakeServerRequest.then(() => {
                  throw new Error('AsyncValidationSample.userName.submit.failed');
              });
    };

    onSubmitFailed = (e?: Error): void => {
        e &&
            this.setState({
                data: {
                    values: { ...this.state.data.values },
                    errors: { ...this.state.data.errors, userName: { id: e.message } },
                },
            });
    };

    onSubmitFinished = (): void => {
        this.setState({ data: { values: {}, errors: {} } });
    };

    render(): React$Node {
        const { data, pendingUserName } = this.state;
        const { values, errors, submitting } = data;
        return (
            <div className="col-12">
                <Form
                    validation={this.state.validations}
                    onChange={this.onChange}
                    data={data}
                    onSubmit={this.onSubmit}
                    onSubmitFailed={this.onSubmitFailed}
                    onSubmitFinished={this.onSubmitFinished}>
                    <div className="row">
                        <div className="col-md-12">
                            <FormInput
                                name="userName"
                                value={values.userName}
                                error={errors.userName}
                                pending={pendingUserName}
                                label="Username"
                                placeholder="Please enter your desired username"
                            />
                            <p className="small font-italic form-group-apply">
                                ATTENTION: The validation succeeds for all names, but the following:<br />
                                <strong>Tom</strong>: Validation fails each time.<br />
                                <strong>Jack</strong>: Validation fails each second time.<br />
                                <strong>Mike</strong>: Validation succeeds but submitting fails.
                            </p>
                            <FormInput
                                name="realName"
                                value={values.realName}
                                error={errors.realName}
                                label="Real name"
                                placeholder="Please enter your real name"
                            />
                            <div className="btn-toolbar">
                                <button className="btn btn-success" disabled={submitting || Form.hasErrors(data)}>
                                    {submitting && <i className="fa fa-circle-notch fa-spin" />} Submit
                                </button>
                            </div>
                        </div>
                    </div>
                </Form>
            </div>
        );
    }
}
