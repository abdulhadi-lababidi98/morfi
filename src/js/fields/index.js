/**
 * This file is part of form4react which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import { _FormInput as FormInput } from './FormInput';

export { FormInput };
