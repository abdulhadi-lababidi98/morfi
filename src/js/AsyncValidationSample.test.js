/**
 * This file is part of form4react which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import React from 'react';
import AsyncValidationSample from './AsyncValidationSample';
import { Spy } from 'spy4js';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Input } from './fields/FormInput';
import { FormWrapper } from '../test/form-test-util';

describe('AsyncValidationSample', () => {
    it('renders the initial form', () => {
        expect(toJson(shallow(<AsyncValidationSample />))).toMatchSnapshot();
    });

    it('validates the form integration', async () => {
        const wrapper = new FormWrapper(<AsyncValidationSample />)
            .validate(
                { base: 'state', name: 'data' },
                { name: 'userName', component: Input, required: true },
                { name: 'realName', component: Input, required: false }
            )
            .lookAt('realName')
            .change('Tester')
            .valueIs('Tester')
            .errorIs(undefined)
            .change('Tester @ga!n')
            .valueIs('Tester @ga!n')
            .errorIs({ id: 'AsyncValidationSample.realName.validation.requirements' })
            .lookAt('userName')
            .change('TestFoo')
            .errorIs(undefined);

        const mathRandomSpy = Spy.on(Math, 'random').returns(0.9);
        await wrapper.blur().nextTick();

        wrapper
            .valueIs('TestFoo')
            .errorIs({ id: 'AsyncValidationSample.userName.already.registered', values: { userName: 'TestFoo' } })
            .change('otherFooName')
            .errorIs(undefined);

        mathRandomSpy.returns(0.1);
        await wrapper.blur().nextTick();

        wrapper.valueIs('otherFooName').errorIs(undefined);
    });
});
