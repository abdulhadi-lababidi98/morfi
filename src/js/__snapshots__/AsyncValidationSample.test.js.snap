// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncValidationSample renders the initial form 1`] = `
<div
  className="col-12"
>
  <_Form
    data={
      Object {
        "errors": Object {},
        "values": Object {},
      }
    }
    onChange={[Function]}
    onSubmit={[Function]}
    onSubmitFailed={[Function]}
    onSubmitFinished={[Function]}
    validation={
      Object {
        "realName": Object {
          "onChange": [Function],
        },
        "userName": Object {
          "onBlur": [Function],
          "onChange": [Function],
        },
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="col-md-12"
      >
        <Component
          label="Username"
          name="userName"
          pending={false}
          placeholder="Please enter your desired username"
        />
        <p
          className="small font-italic form-group-apply"
        >
          ATTENTION: The validation succeeds for all names, but the following:
          <br />
          <strong>
            Tom
          </strong>
          : Validation fails each time.
          <br />
          <strong>
            Jack
          </strong>
          : Validation fails each second time.
          <br />
          <strong>
            Mike
          </strong>
          : Validation succeeds but submitting fails.
        </p>
        <Component
          label="Real name"
          name="realName"
          placeholder="Please enter your real name"
        />
        <div
          className="btn-toolbar"
        >
          <button
            className="btn btn-success"
            disabled={false}
          >
             Submit
          </button>
        </div>
      </div>
    </div>
  </_Form>
</div>
`;
